### PR TITLE
Improve keyboard interaction of the DropdownMenu

### LINKED ISSUE #1880: DropdownMenu a11y improvements and proposal to reuse it for other UI controls
See for example #1879 

The `DropdownMenu` component is pretty good for accessibility and works pretty well, could also be further improved.
- it already uses `role="menu"` and  `role="menuitem"`
- already implements arrows navigation
- it gets its items via props

Currently, it's used just for the Table Block:

screen shot 2017-07-13 at 11 56 13

Wondering if it could be used also in other places, for example:
- the block switcher (it's a menu and works in a similar way)
- the format list (same, even if it has some peculiarities)

Further improvements to consider:
- navigation should be implemented with arrows only (no Tab key!) and it's described in detail on https://www.w3.org/TR/wai-aria-practices/#menu

While the recommended pattern described above is a bit hard to follow, I'd strongly recommend to dedicate a few minutes to it to get how interaction should work.

The advantage of having just arrows navigation is that the amount of Tab key presses required for navigation is drastically reduced: users would have the choice to navigate through the main UI components with the Tab key and enter the menu with arrows key.

To get an idea, the ARIA Authoring Practices have also an example, I'd recommend to try the interaction on this menubar example:
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html

Worth noting the DropdownMenu currently uses a keydown event also for the Tab key, to move focus to the previous/next items. Not sure that's really needed, since the items are buttons so they're natively focusable and already in a correct Tab order. Regardless, the ARIA menu pattern requires just arrows navigation. The Tab key should be used to exit the menu.

TL;DR
More details on the expected keyboard interaction:

https://www.w3.org/TR/wai-aria-practices/#menu
> When a menu opens, or when a menubar receives focus, keyboard focus is placed on the first item. All items are focusable as described in 4.6 Keyboard Navigation Inside Components.

4.6 Keyboard Navigation Inside Components
https://www.w3.org/TR/wai-aria-practices/#kbd_general_within
> As described in section 4.1 Fundamental Keyboard Navigation Conventions, the tab sequence should include only one focusable element of a composite UI component. Once a composite contains focus, keys other than Tab and Shift + Tab enable the user to move focus among its focusable elements. Authors are free to choose which keys move focus inside of a composite, but they are strongly advised to use the same key bindings as similar components in common GUI operating systems

### PR DESCRIPTION
**Pull request by afercia**
This PR tries to improve the accessibility and keyboard interaction of the DropdownMenu following the model described in the ARIA Authoring Practices.

The DropdownMenu is, in ARIA terms, a composite widget made of a [Menu button](https://www.w3.org/TR/wai-aria-practices/#menubutton) and a [Menu](https://www.w3.org/TR/wai-aria-practices/#menu). To get how the expected keyboard interaction should work I'd recommend to jump directly to the W3C examples for the standalone Menu:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html
and the same example in the context of a Toolbar:
https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html

Navigation through the menu items should work with arrow keys only, while tabbing should close the menu (same for Escape). Optionally, items navigation should "loop" continuously. 

Worth reminding ARIA widgets replicate common, established patterns already used in our operating systems and keyboard users are used to them. For example, the macOS Finder menu works this way: it uses arrow keys to navigate the drop-down items:

screen shot 2017-07-21 at 19 17 36

The mixed use of the Tab key and arrow keys is a great way to reduce the amount of Tab key presses necessary to navigate through content, as it allows to open a menu just if that's really needed.

As mentioned in the issue, it would be great to consider to extend this behavior to other components too.

Fixes #1880

### CODE REVIEW

**Comment by afercia** [comment]
Fixes #1880

File: components/dropdown-menu/index.js [hunk 1]
```diff
@@ -60,47 +79,58 @@ class DropdownMenu extends wp.element.Component {
...
 			}
 			return -1;
 		}
 	}
 
 	focusIndex( index ) {
-		if ( this.menuRef ) {
-			const menu = findDOMNode( this.menuRef );
+		if ( this.menuNode ) {
 			if ( index < 0 ) {
```
Comment (aduth) [inline components/dropdown-menu/index.js]: With changes to `focusPrevious`, is it ever the case that this condition would pass? Also generally concerned about `previousElementSibling` potentially not being a focusable element. Might be worth dropping this.

File: components/dropdown-menu/index.js [hunk 2]
```diff
@@ -47,10 +50,26 @@ class DropdownMenu extends wp.element.Component {
...
+	handleFocus() {
+		this.hasFocus = true;
+	}
+
+	handleBlur() {
+		this.hasFocus = false;
+
+		this.maybeCloseMenuOnBlur = setTimeout( () => {
+			this.closeMenu();
+		}, 100 );
```
Comment (aduth) [inline components/dropdown-menu/index.js]: How did you settle on `100` ?

File: components/dropdown-menu/index.js [hunk 3]
```diff
@@ -60,47 +79,58 @@ class DropdownMenu extends wp.element.Component {
...
 	handleKeyDown( keydown ) {
 		if ( this.state.open ) {
 			switch ( keydown.keyCode ) {
 				case ESCAPE:
 					keydown.preventDefault();
 					keydown.stopPropagation();
-					this.closeMenu();
 					const node = findDOMNode( this );
 					const toggle = node.querySelector( '.components-dropdown-menu__toggle' );
 					toggle.focus();
```
Comment (aduth) [inline components/dropdown-menu/index.js]: I'm finding that occasionally pressing escape when navigating buttons will clear focus altogether (`document.activeElement === document.body`). Not sure if that's specific to your changes here though.

macOS Chrome 59.0.3071.115

File: components/dropdown-menu/index.js [hunk 4]
```diff
@@ -192,7 +219,10 @@ class DropdownMenu extends wp.element.Component {
 						className="components-dropdown-menu__menu"
 						role="menu"
 						aria-label={ menuLabel }
-						ref={ this.bindMenuRef }
+						ref={ this.bindMenuReferenceNode }
+						onFocus={ this.handleFocus }
+						onBlur={ this.handleBlur }
+						tabIndex="-1"
```
Comment (aduth) [inline components/dropdown-menu/index.js]: Is `tabindex` necessary on `div`?

File: components/dropdown-menu/index.js (hunk 2 above)
Comment (afercia) [inline components/dropdown-menu/index.js]: Completely arbitrary (I know...). There's a brief interval between blur on an element and focus on the following element. '0' didn't work. On the W3C examples, they use '300', see:
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/Menubutton.js
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/MenuItemAction.js
https://www.w3.org/TR/wai-aria-practices/examples/menu-button/js/PopupMenuAction.js

Open to suggestions...

File: components/dropdown-menu/index.js (hunk 1 above)
Comment (afercia) [inline components/dropdown-menu/index.js]: OK

File: components/dropdown-menu/index.js (hunk 3 above)
Comment (afercia) [inline components/dropdown-menu/index.js]: Same chrome here and can't reproduce. Updated to 60: same.
Instead, I think `handleKeyUp()` doesn't work as intended: when I keep Escape pressed for a while, the menu closes immediately so when I release Escape the condition to stop propagation isn't met and the toolbar disappears.

**Comment by afercia** [comment]
Using keyup to close the menu when pressing Escape allows to greatly simplify. No timers, no focus/blur. To recap:
- menu items can be navigated with arrows only
- tabbing will close the menu and natively focus the next/previous focusable elements (note: there's a pre-existing and unrelated bug when using Firefox, will open a separate issue)
- pressing Escape closes the menu and focuses the toggle
- one more Escape press makes the whole toolbar disappear, as expected

File: components/dropdown-menu/index.js [hunk 5]
```diff
@@ -60,62 +61,62 @@ class DropdownMenu extends wp.element.Component {
...
+		/*
+		 * VisualEditorBlock uses a keyup event to deselect the block. When the
+		 * menu is open we need to stop propagation after Escape has been pressed
+		 * so we use a keyup event instead of keydown, otherwise the whole block
+		 * toolbar will disappear.
+		 */
+		if ( event.keyCode === ESCAPE && this.state.open ) {
+			event.preventDefault();
+			event.stopPropagation();
+			const node = findDOMNode( this );
```
Comment (aduth) [inline components/dropdown-menu/index.js]: We could probably create a separate `ref` for the root rendered element to avoid the `findDOMNode` here.

File: components/dropdown-menu/index.js [hunk 6]
```diff
@@ -60,62 +61,62 @@ class DropdownMenu extends wp.element.Component {
...
+			const toggle = node.querySelector( '.components-dropdown-menu__toggle' );
```
Comment (aduth) [inline components/dropdown-menu/index.js]: Actually, for that matter, we could bind the `ref` to the toggle to avoid the `querySelector` as well.

**Comment by aduth** [comment]
Would be good to have some unit tests for this component, but that can be addressed separately I think.

**Comment by aduth** [comment]
In e91f3a2, I revised `IconButton` to be a stateful component class, so that we can apply a `ref` to target it directly shifting focus to the toggle when the menu closes. Let me know if this looks good to you, otherwise I think this is ready to merge.

**Review by aduth — APPROVED** [review]
